third_party/zstd-jni/zstd-jni_1.4.5-6.patch

```diff
@@ -0,0 +1,227 @@
+     }
+ 
+     public static synchronized void load() {
+-        load(null);
+-    }
+-
+-    public static synchronized void load(final File tempFolder) {
+         if (loaded) {
+             return;
+         }
+-
+-        String resourceName = resourceName();
+-        InputStream is = Native.class.getResourceAsStream(resourceName);
+-        if (is == null) {
+-            // fall-back to loading the zstd-jni from the system library path.
+-            // It also cover loading on Android.
+-            try {
+-                System.loadLibrary(libnameShort);
+-                loaded = true;
+-                return;
+-            } catch (UnsatisfiedLinkError e) {
+-                UnsatisfiedLinkError err = new UnsatisfiedLinkError(e.getMessage() + "\n" + errorMsg);
+-                err.setStackTrace(e.getStackTrace());
+-                throw err;
+-            }
++        try {
++            System.loadLibrary(libnameShort);
++        } catch (UnsatisfiedLinkError e) {
++            // Try to load from runfiles, esp. for Windows.
++            loadFromRunfiles();
+         }
+-        File tempLib = null;
+-        FileOutputStream out = null;
++    }
++
++    private static synchronized void loadFromRunfiles() {
++        Runfiles runfiles = null;
+         try {
+-            tempLib = File.createTempFile(libname, "." + libExtension(), tempFolder);
+-            // try to delete on exit, does not work on Windows
+-            tempLib.deleteOnExit();
+-            // copy to tempLib
+-            out = new FileOutputStream(tempLib);
+-            byte[] buf = new byte[4096];
+-            while (true) {
+-                int read = is.read(buf);
+-                if (read == -1) {
+-                    break;
+-                }
+-                out.write(buf, 0, read);
+-            }
+-            try {
+-                out.flush();
+-                out.close();
+-                out = null;
+-            } catch (IOException e) {
+-                // ignore
+-            }
+-            try {
+-                System.load(tempLib.getAbsolutePath());
+-            } catch (UnsatisfiedLinkError e) {
+-                // fall-back to loading the zstd-jni from the system library path
+-                try {
+-                    System.loadLibrary(libnameShort);
+-                } catch (UnsatisfiedLinkError e1) {
+-                    // display error in case problem with loading from temp folder
+-                    // and from system library path - concatenate both messages
+-                    UnsatisfiedLinkError err = new UnsatisfiedLinkError(
+-                            e.getMessage() + "\n" +
+-                            e1.getMessage() + "\n"+
+-                            errorMsg);
+-                    err.setStackTrace(e1.getStackTrace());
+-                    throw err;
+-                }
+-            }
+-            loaded = true;
++            runfiles = Runfiles.create();
```

Comment:
Why do you use the Runfiles library to find the libzstd-jni.dll? It is shipped within the Bazel binary, right?

Reply:
This is how we load JNI library in Bazel:
https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/jni/JniLoader.java;l=54?q=%22java.library.path%22&ss=bazel

I think here: `System.loadLibrary("libzstd-jni");` should just work.

Reply:
This code was based on the way Bazel did it before https://github.com/bazelbuild/bazel/commit/a2a2f9d5f6b that was commited 9 days ago. I'll try to unify it with current version (hopefully no need to create files), and test it under windows (setting up VM right now...) hopefully early next week.